nexus/src/authz/api_resources.rs

```diff
@@ -19,13 +19,87 @@
 // pass them to `is_allowed()`.  Using newtypes is a way to capture just the
 // parts we need for authorization.
 
-use super::actor::AuthenticatedActor;
-use super::roles::AuthzApiResource;
-use super::roles::AuthzResource;
+use super::actor::AnyActor;
+use super::context::Authorize;
+use super::roles::{
+    load_roles_for_resource, load_roles_for_resource_tree, RoleSet,
+};
+use super::Action;
+use super::{actor::AuthenticatedActor, Authz};
+use crate::authn;
+use crate::context::OpContext;
 use crate::db::fixed_data::FLEET_ID;
-use omicron_common::api::external::ResourceType;
+use crate::db::DataStore;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use omicron_common::api::external::{Error, LookupType, ResourceType};
 use uuid::Uuid;
 
+/// Describes an authz resource that corresponds to an API resource that has a
+/// corresponding ResourceType and is stored in the database
+///
+/// This is a helper trait used to impl [`AuthzResource`].
+pub trait AuthzApiResource: Clone + Send + Sync + 'static {
+    /// If roles can be assigned to this resource, return the type and id of the
+    /// database record describing this resource
+    ///
+    /// If roles cannot be assigned to this resource, returns `None`.
+    fn db_resource(&self) -> Option<(ResourceType, Uuid)>;
+
+    /// If this resource has a parent in the API hierarchy whose assigned roles
+    /// can affect access to this resource, return the parent resource.
+    /// Otherwise, returns `None`.
+    fn parent(&self) -> Option<&dyn Authorize>;
+
+    /// Returns an error as though this resource were not found, suitable for
+    /// use when an actor should not be able to see that this resource exists
+    fn not_found(&self) -> Error;
+}
+
+impl<T: AuthzApiResource + oso::PolarClass> Authorize for T {
+    fn load_roles<'a, 'b, 'c, 'd, 'e, 'f>(
+        &'a self,
+        opctx: &'b OpContext,
+        datastore: &'c DataStore,
+        authn: &'d authn::Context,
+        roleset: &'e mut RoleSet,
+    ) -> BoxFuture<'f, Result<(), Error>>
+    where
+        'a: 'f,
+        'b: 'f,
+        'c: 'f,
+        'd: 'f,
+        'e: 'f,
+    {
+        load_roles_for_resource_tree(self, opctx, datastore, authn, roleset)
+            .boxed()
+    }
+
+    fn on_unauthorized(
+        &self,
+        authz: &Authz,
+        error: Error,
+        actor: AnyActor,
+        action: Action,
+    ) -> Result<(), Error> {
+        if action == Action::Read {
```

Comment:
This function is the crux of the change.  It's invoked by the main `authorize()` function if authorization fails.  Here we're saying if we were already checking for "read", and that failed, then return 404.  Otherwise, do a "read" check and return 404 (if that fails) or whatever error `authorize()` had already picked (which is either 401 or 403 based on whether the user was authenticated).